### ISSUE TITLE
Using matplotlib.subplot with GeoAxes fails to give same size panels

### ISSUE THREAD

**Issue by alimoridnejad** [matplotlib/matplotlib#11374]
### Bug report

**Bug summary**
A bug in the subclass of matplotlib.axes.Axes called GeoAxes. When doing subplot, it often scales the size of last panel differently compared to the rest of panels. The issue arises only when using matplotlib.subplot and does not appear for single plot applications. Therefore to avoid this bug, one must always play around with figsize argument in subplot and axis.set_extent to make the required adjustments. I think it would be nicer to fix this bug rather than spending so much time trying to justify the wrong scaling in the plot.

**Code for reproduction**

```python
fig, axes = plt.subplots(nrows=3, ncols=2,
                         figsize=(8,15), 
                         sharex=True, sharey=True, 
                         subplot_kw = dict(projection=ccrs.PlateCarree()))
for i in range(6):
    axes.flat[i].set_extent([-70, -30, 35, 70])
    fill = axes.flat[i].contourf(lons, lats, eof[i].squeeze(),
                                 clevs,
                                 transform = ccrs.PlateCarree(),
                                 cmap = plt.cm.jet,
                                 extend='both')
```

**Actual outcome**
[test.pdf](https://github.com/matplotlib/matplotlib/files/2070103/test.pdf)

**Expected outcome**
Size of the last panel should be equal to the size of all other panels in a subplot  

**Matplotlib version**

  * Operating system: Linux Ubuntu 16.4 
  * Matplotlib version: 2.2.2
  * Matplotlib backend: nbAgg
  * Python version: 3.6.5
  * Jupyter version: 5.4.1
  * Other libraries: Cartopy 0.16.0

 Conda

**Comment by ImportanceOfBeingErnest** [comment]
Any chance for a minimal, **complete** (i.e. runnable) example?

**Comment by alimoridnejad** [comment]
I run the code on a big data that I can't attach here. But here is the full code: 
[Code_with_example.txt](https://github.com/matplotlib/matplotlib/files/2072745/Code_with_example.txt)

Let me know if you have questions. Also, the code shows the output figure after the second run. Any ideas why it does not show it in the first run?

**Comment by ImportanceOfBeingErnest** [comment]
Doesn't seem like this is a minimal, complete example. It also seems that the issue is rather with cartopy than with matplotlib. So in total I'm not convinced this is an issue one can solve here.

**Comment by alimoridnejad** [comment]
I think it is a bit soon to tag this as a cartopy issue, although I will consider it for later tests. I cut the code to provide a minimal complete example. 
[Minimal_example.pdf](https://github.com/matplotlib/matplotlib/files/2077364/Minimal_example.pdf)

I tested many examples to better understand this bug. I just forked and clone the matplotlib and will start working on this issue. I'd like to start looking at different classes of axes now and learn how they are linked to a projection system from other matplotlib toolkit.

**Comment by alimoridnejad** [comment]
Thanks for your feedback. Sorry I am new to github and I thought by minimal you meant a short code only. I'd like to learn how make a reproducible version. The data I have are in the network of common data form (netCDF format). Can you let me know what would people do when they pickle from a large dataset and send it to others?

**Comment by alimoridnejad** [comment]
Thanks for your feedback. Sorry I am new to github and I thought by minimal you meant a short code only. I'd like to learn how make a reproducible version. The data I have are in the network of common data form (netCDF format). Can you let me know what would people do when they pickle from a large dataset and send it to others? Or any guide on making this code reproducable?

**Comment by dopplershift** [comment]
If you set the extents of the subplots manually (using `set_extent`), I suspect you can reproduce
without needing to plot the original data.

**Comment by alimoridnejad** [comment]
Thanks Ryan! I am going to try it now :)

**Comment by gepcel** [comment]
I've ran into some same issue long ago (with the 1st subplot), too long to recall. Not even sure that issue still exists.

But regard to your example. There is something you should try. You set the extent of the axes to `.set_extent([-70, -30, 35, 70]`, and then set the yticks to `.set_yticks([30, 40, 50, 60, 70]`, which is out of the extent boundary.

**Comment by afvincent** [comment]
@alimoridnejad Because time is a scarce ressource for most people, an issue ticket will be *even more* (;-)) appreciated if it contains an “S(S)SCCE”, which stands for:

* Short (Small) - Minimize bandwidth for your example, do not bore the audience.
* Self Contained - Ensure everything is included, ready to go.
* Correct - Copy, paste, (compile,) see is the aim.
* Example - Displays the problem you are trying to solve.

(shamelessly copy-pasted from [http://www.sscce.org/](http://www.sscce.org/) that @jklymak likes to use :sheep:)

In the case of Matplotlib, such an example would for example try to minimize the amount of imports. Most of the time, you can cut it down to just `matplotlib`, and possibly `numpy` to generate (as simple as possible) artificial data that trigger to issue (this way the devs are sure that the is likely to come from Matplotlib). If one really has to use original data, those have to be provided as well, under a format that is easy to manipulate (e.g. a short CSV file, possibly extracted from a larger data set). Besides the length of the whole example should be as short as possible to make it easier to circumvent the possible origin of the issue (e.g. do not use aesthetic parameters if they are not useful for trigerring the issue). In the end, *the goal is to make it easy for other people to copy-paste-and-run your example in order to reproduce the issue on their machine.*

**Comment by ImportanceOfBeingErnest** [comment]
The main point being reproducibility really. If a code is not minimal, but complete and runnable, that may be annoying for people but is no blocker for tracking down the issue. If a code is minimal, but not complete, there is really no chance of seeing if there is an issue at all. 

> what would people do when they pickle from a large dataset and send it to others?

That should almost never happen for issues reported here. You should always be able to replace your large dataset by some data you produce programmatically within the code you share.

**Comment by alimoridnejad** [comment]
The data and the code in Jupyter Notebook required to reproduce the error is attached here: 
[runnable_version.zip](https://github.com/matplotlib/matplotlib/files/2082279/runnable_version.zip)

**Comment by alimoridnejad** [comment]
@dopplershift I ended up using `numpy` objects to extract the data instead of using `set_extent`. But as you pointed out in your talk at SciPy 2017, sometimes `numpy` decides you don't need units and I could not see units when creating these objects. Do you have any thoughts on this issue or has been any progress to keep the units when extracting/converting the data? Thanks

**Comment by ImportanceOfBeingErnest** [comment]
That is what people would call annoying. Because as it turns out one of the above comments was absolutly right, and you can simply set the extent to reproduce, without any data. 

A minimal complete example hence looks like this:

```
import cartopy.crs as ccrs
import numpy as np
import matplotlib.pyplot as plt
%matplotlib notebook

fig, axes = plt.subplots(nrows=3, ncols=2,
                         figsize=(8,6),
                         sharex=True, sharey=True, 
                         subplot_kw = dict(projection=ccrs.PlateCarree()))

for i in range(6):
    axes.flat[i].set_extent([-70, -30, 35, 70], ccrs.PlateCarree())
    axes.flat[i].set_xticks([-70, -60, -50, -40, -30], crs=ccrs.PlateCarree())
    axes.flat[i].set_yticks([30, 40, 50, 60, 70], crs=ccrs.PlateCarree())
```

and you can solve the issue by not setting any ticks outside the extent , e.g. replace the number 35 by 30, 
`[-70, -30, 30, 70]`, or replace the number 30 by 35, `[35, 40, 50, 60, 70]`. Exactly as @gepcel commented above already.

I would say this *is* a cartopy issue, because it depends on `set_extent` which is a cartopy function. So please report this with cartopy and not with matplotlib.

**Comment by alimoridnejad** [comment]
@gepcel Yes, it sometimes changes the size of the first panel on the subplot instead of last one even if plot extent are identical to the axis limits for labelling. I wanted to highlight the issue here by changing the labels here. Also, It should not change the aspect ratio of one plot only. It is really just about the labels.

**Comment by ImportanceOfBeingErnest** [comment]
As said, please report with cartopy and make sure not to annoy them with a non-minimal example as well.

**Comment by alimoridnejad** [comment]
@ImportanceOfBeingErnest Thank you Elan. I will discuss this with the `cartopy` community.
